zerver/views/custom_profile_fields.py

```diff
@@ -0,0 +1,74 @@
+                                do_remove_realm_custom_profile_field,
+                                do_update_realm_custom_profile_field,
+                                do_update_user_custom_profile_data)
+from zerver.lib.response import json_success, json_error
+from zerver.lib.validator import check_dict, check_list, check_int, check_string
+
+from zerver.models import (custom_profile_fields_for_realm, UserProfile,
+                           CustomProfileField, custom_profile_fields_for_realm)
+
+def list_realm_custom_profile_fields(request, user_profile):
+    # type: (HttpRequest, UserProfile) -> HttpResponse
+    fields = custom_profile_fields_for_realm(user_profile.realm_id)
+    return json_success({'custom_fields': fields})
+
+@require_realm_admin
+@has_request_variables
+def create_realm_custom_profile_field(request, user_profile, name=REQ(),
+                                      field_type=REQ(validator=check_int)):
+    # type: (HttpRequest, UserProfile, Text, int) -> HttpResponse
+    try:
+        field_id = do_add_realm_custom_profile_field(
+            realm=user_profile.realm,
+            name=name,
+            field_type=field_type,
+        )
+        return json_success({'id': field_id})
+    except ValidationError as e:
+        return json_error(e.messages[0], data={"errors": dict(e)})
+
+@require_realm_admin
+def delete_realm_custom_profile_field(request, user_profile, field_id):
+    # type: (HttpRequest, UserProfile, int) -> HttpResponse
+    try:
+        do_remove_realm_custom_profile_field(realm=user_profile.realm,
+                                             field_id=field_id)
+    except CustomProfileField.DoesNotExist:
+        return json_error(_('Field not found'))
+    return json_success()
+
+@require_realm_admin
+@has_request_variables
+def update_realm_custom_profile_field(request, user_profile, field_id,
+                                      name=REQ(),
+                                      field_type=REQ(validator=check_int)):
+    # type: (HttpRequest, UserProfile, int, Text, int) -> HttpResponse
+    try:
+        do_update_realm_custom_profile_field(user_profile.realm,
+                                             field_id, name, field_type)
+    except CustomProfileField.DoesNotExist:
+        return json_error(_('Field not found'))
+    except IntegrityError:
+        return json_error(_('Name of the field should be unique in a realm.'))
+    return json_success()
+
+@has_request_variables
+def update_user_custom_profile_data(
+        request,
+        user_profile,
+        data=REQ(validator=check_list(check_dict([('id', check_int),
+                                                  ('value', check_string)])))):
+    # type: (HttpRequest, UserProfile, List[Dict[str, Union[int, Text]]]) -> HttpResponse
+    do_update_user_custom_profile_data(user_profile, data)
+    return json_success({'fields': [field['id'] for field in data]})
```

Comment:
Stylistically, where there isn't a significant race issue, we generally prefer to do validation for whether the input is valid (e.g. the field referenced exists, or the value isn't already there, or whatever) in the view function layer, not at the `do_update_...` layer; functions with names starting with `do_` should just execute the operation with input that is already validated.  I think you could either rename these to e.g. `check_` or `try_` for clarity, or move the input validation into the view layer (which might be a better model anyway).  

Certainly this function seems like it should be checking if the fields are valid for the given realm here, since you don't have that check anywhere else.